Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm

```diff
@@ -28,25 +28,100 @@
 #endif
 
 #import "config.h"
+#import "MessageSenderInlines.h"
 #import "WebExtensionAPINamespace.h"
 #import "WebExtensionAPIWebNavigation.h"
+#import "WebExtensionContext.h"
+#import "WebExtensionContextMessages.h"
+#import "WebExtensionFrameIdentifier.h"
+#import "WebExtensionFrameParameters.h"
+#import "WebExtensionUtilities.h"
+#import "WebProcess.h"
 
 #if ENABLE(WK_WEB_EXTENSIONS)
 
 namespace WebKit {
 
+static NSString *tabIdKey = @"tabId";
+static NSString *frameIdKey = @"frameId";
+static NSString *parentFrameIdKey = @"parentFrameId";
+static NSString *errorOccurredKey = @"errorOccurred";
+static NSString *urlKey = @"url";
+
+static NSDictionary *toWebAPI(const WebExtensionFrameParameters& frameInfo)
+{
+    NSMutableDictionary *result = [NSMutableDictionary dictionary];
+
+    result[errorOccurredKey] = frameInfo.errorOccurred ? @(frameInfo.errorOccurred.value()) : @NO;
+    result[parentFrameIdKey] = @(toWebAPI(frameInfo.parentFrameIdentifier));
+    result[urlKey] = frameInfo.url ? (NSString *)frameInfo.url.value().string() : @"";
```

Comment:
i think alternatively this could work?
`result[urlKey] = (NSString *)frameInfo.url.value_or(""_s)`

Reply:
In this case it would not work, since `value()` is `URL`.

Tabs does this, to account for null URLs so they are returned as empty strings too.

```
if (parameters.url)
    result[urlKey] = !parameters.url.value().isNull() ? (NSString *)parameters.url.value().string() : emptyURLValue;
```

The `url` key is optional in tabs, so you might need to format differently to handle the optional and isNull on URL if `url` is a required key for frame info.

Like this:

```
result[urlKey] = parameters.url && !parameters.url.value().isNull() ? (NSString *)parameters.url.value().string() : emptyURLValue;
```

Reply:
I think the current structure is a bit more readable, but don't feel strongly about it.